### ISSUE TITLE
F-Droid build failed

### ISSUE THREAD

**Issue by linsui** [tutao/tutanota#3154]
- [x] This is not a feature request (existing functionality does not work, **not** missing functionality).
 I will request features on [forum](https://www.reddit.com/r/tutanota/) or via support.
- [x] I've searched and did not find a similar issue.

**Bug in mobile app**

**Describe the bug**
```
+ node dist prod
internal/process/esm_loader.js:74
    internalBinding('errors').triggerUncaughtException(
                              ^

Error [ERR_MODULE_NOT_FOUND]: Cannot find package 'node-forge' imported from /home/vagrant/build/de.tutao.tutanota/buildSrc/installerSigner.js
    at packageResolve (internal/modules/esm/resolve.js:655:9)
    at moduleResolve (internal/modules/esm/resolve.js:696:18)
    at Loader.defaultResolve [as _resolve] (internal/modules/esm/resolve.js:810:11)
    at Loader.resolve (internal/modules/esm/loader.js:86:40)
    at Loader.getModuleJob (internal/modules/esm/loader.js:230:28)
    at ModuleWrap.<anonymous> (internal/modules/esm/module_job.js:56:40)
    at link (internal/modules/esm/module_job.js:55:36) {
  code: 'ERR_MODULE_NOT_FOUND'
}
```

Could you please take a look?

**Comment by jowlo** [comment]
We remove `node-froge` in `fix-fdroid-deps` to remove dependecies with blobs, but `installerSigner` needs `node-forge` since f90e3f80abf0741df4e8e7b699884d441c086d17 for custom certs. 

Since F-Droid will not use this functionality anyways, could we do an async import of `node-forge` here @ganthern ?
https://github.com/tutao/tutanota/blob/master/buildSrc/installerSigner.js#L94-L99

**Comment by jowlo** [comment]
So, after spending *way, way* too much time on trying to set up the F-Droid build server environment and failing in all kinds of possible ways, could you @linsui check that running the F-Droid build against the `fix-fdroid-build-3154` branch works as expected?

If so we can do an android release to fix the build on monday. Thank you!

**Comment by linsui** [comment]
You can fork the [fdroid/fdroiddata](https://gitlab.com/fdroid/fdroiddata) repo and use the CI directly by pushing your change. I'll check it tomorrow. Thanks!

**Comment by linsui** [comment]
When executing `+ node dist prod` it failed. See https://gitlab.com/linsui/fdroiddata/-/jobs/1321251064.

```
/src/calendar/date/AlarmScheduler.js
/src/calendar/date/CalendarInvites.js
jszip-57e67c20.js  94.55078125K
libs/jszip.js
If you do not supply "output.name", you may not be able to access the exports of an IIFE bundle.
Build error: Error [ERR_MODULE_NOT_FOUND]: Cannot find package 'node-forge' imported from /builds/linsui/fdroiddata/build/de.tutao.tutanota/buildSrc/installerSigner.js
    at packageResolve (internal/modules/esm/resolve.js:655:9)
    at moduleResolve (internal/modules/esm/resolve.js:696:18)
    at Loader.defaultResolve [as _resolve] (internal/modules/esm/resolve.js:810:11)
    at Loader.resolve (internal/modules/esm/loader.js:86:40)
    at Loader.getModuleJob (internal/modules/esm/loader.js:230:28)
    at ModuleWrap.<anonymous> (internal/modules/esm/module_job.js:56:40)
    at link (internal/modules/esm/module_job.js:55:36) {
  code: 'ERR_MODULE_NOT_FOUND'
}
2021-06-05 11:23:21,773 ERROR: Could not build app de.tutao.tutanota: Error running prebuild command for de.tutao.tutanota:3.84.1
```

**Comment by jowlo** [comment]
Thank you @linsui for testing this for us. I forked the repo and am testing another fix at the moment.

We have a working local setup now as well and are planning to incorporate the F-Droid build into our own build server so we get notified of breaking changes.

We will prepare another android release to be built by F-Droid soon.

**Comment by linsui** [comment]
Thanks!

**Comment by linsui** [comment]
```
If you do not supply "output.name", you may not be able to access the exports of an IIFE bundle.
Build error: Error [ERR_MODULE_NOT_FOUND]: Cannot find package 'node-forge' imported from /home/vagrant/build/de.tutao.tutanota/buildSrc/installerSigner.js
    at packageResolve (internal/modules/esm/resolve.js:650:9)
    at moduleResolve (internal/modules/esm/resolve.js:691:18)
    at Loader.defaultResolve [as _resolve] (internal/modules/esm/resolve.js:805:11)
    at Loader.resolve (internal/modules/esm/loader.js:88:40)
    at Loader.getModuleJob (internal/modules/esm/loader.js:241:28)
    at ModuleWrap.<anonymous> (internal/modules/esm/module_job.js:72:40)
    at link (internal/modules/esm/module_job.js:71:36) {
  code: 'ERR_MODULE_NOT_FOUND'
}
```

3.84.8 failed. Could you please take a look? Thanks!

**Comment by vitoreiji** [comment]
@linsui can you let us know if 3.85.4 is building OK?

I had a look at our [metdata file](https://gitlab.com/fdroid/fdroiddata/-/blob/master/metadata/de.tutao.tutanota.yml) but that seems to only have up to 3.85.3, and I could not find the corresponding pipeline to see if there was a build problem with that version (is there an easy way to find this?)

**Comment by jowlo** [comment]
@vitoreiji you can always just fork the metadata repository, make changes (eg include the next version) and trigger the build on the same build pool automatically with a push.

I also couldn't find the pipeline that ran the last version. neither do i know why the checkupdates bot did not pick up the new version. We are listed here, but i am unsure what it means:  https://monitor.f-droid.org/builds/needsupdate

**Comment by linsui** [comment]
The update checker ran about 8 hours ago but 3.85.4 is released 6 hours ago so it hasn't been added.

**Comment by vitoreiji** [comment]
Build for 3.85.4 is no good: https://gitlab.com/vitoreiji/fdroiddata/-/jobs/1438564046

I suspect the same has been true for all recent versions. Since the error message seems to be the same, I'm reopening this issue.

**Comment by vitoreiji** [comment]
I'm also removing the `not-testable` tag.

# Test notes

1. Fork [fdroid/fdroiddata](https://gitlab.com/fdroid/fdroiddata) on gitlab.com.
2. On that fork, update `metadata/de.tutao.tutanota.yml` to point to the tutanota version containing the fix.
3. Push your changes to your fork go to "CI/CD" > "Jobs" to see the result.

**Comment by linsui** [comment]
Yep, 3.85.3 also failed. https://monitor.f-droid.org/builds/log/de.tutao.tutanota/385030

**Comment by bedhub** [comment]
The build fails because the npm module `node-forge ` is not present but when require but it is declared as dev dependency in our package.json.  Together with @rosso-ptg we checked the output and noticed that during the f-droid build we execute the script `node fdroid-fix-deps.js` which removes the dependency to node-forge (strange that we have to do this). So it seems that this dependency is not necessary for android builds. We checked our `dist.js` for what this dependency is good for and we noticed that it is used when signing desktop client builds and for some reason we call `signDesktopClients()` for every build. The function itself uses and imports `installerSigner.js `which imports` node-forge`.

I could not clarify why it is not done together with buildDesktopClient because then the build should be save as we only import the module when the function is executed.
https://github.com/tutao/tutanota/blob/85301639c8ff37331e431bc74eb5ac003746b97e/dist.js#L131-L137

Initially i wanted to move the function call into the if condition but then we noticed from the git history that it was there before and it was moved out on purpose. As today there is nobody available to clarify the question completely i introduced a quick fix which just delays the import statement a little bit more inside 'signDesktopClients()` just before the actual signing. See commit 8530163
